model/lifecycle.go

```diff
@@ -55,25 +55,79 @@ type VersionToRestart struct {
+	}
+	changed, err := setTaskActivationForVersion(versionId, active, caller)
 	if err != nil {
-		return errors.Wrapf(err, "getting builds for version '%s'", versionId)
+		return errors.Wrapf(err, "setting activation for tasks in version '%s'", versionId)
 	}
-	buildIDs := make([]string, 0, len(builds))
-	for _, build := range builds {
-		buildIDs = append(buildIDs, build.Id)
+	if active && changed && evergreen.IsFinishedVersionStatus(version.Status) {
+		if err = version.UpdateStatus(evergreen.VersionCreated); err != nil {
+			return errors.Wrapf(err, "setting status for version '%s'", versionId)
+		}
 	}
+	return nil
+}
 
-	// Update activation for all builds before updating their tasks so the version won't spend
-	// time in an intermediate state where only some builds are updated
-	if err = build.UpdateActivation(buildIDs, active, caller); err != nil {
-		return errors.Wrapf(err, "setting activation for builds in version '%s'", versionId)
+// setTaskActivationForVersion activates all tasks in a version.
+// Returns true if at least one task has been updated.
+func setTaskActivationForVersion(versionId string, active bool, caller string) (bool, error) {
+	q := bson.M{
+		task.VersionKey: versionId,
+		task.StatusKey:  evergreen.TaskUndispatched,
+	}
+	// If activating a task, set the ActivatedBy field to be the caller.
+	if active {
+		tasksToActivate, err := task.FindAll(db.Query(q).WithFields(task.IdKey, task.DependsOnKey, task.ExecutionKey, task.BuildIdKey))
+		if err != nil {
+			return false, errors.Wrap(err, "getting tasks to activate")
+		}
+		if len(tasksToActivate) > 0 {
+			if err = task.ActivateTasks(tasksToActivate, time.Now(), false, caller); err != nil {
+				return false, errors.Wrap(err, "updating tasks for activation")
+			}
+			var buildIds []string
+			for _, task := range tasksToActivate {
+				if !utility.StringSliceContains(buildIds, task.BuildId) {
+					buildIds = append(buildIds, task.BuildId)
+				}
+			}
+			builds, err := build.Find(build.ByIds(buildIds))
+			if err != nil {
+				return false, errors.Wrap(err, "fetching builds")
+			}
+			for _, b := range builds {
+				_, err := updateBuildStatus(&b)
+				if err != nil {
+					return false, errors.Wrapf(err, "updating status for build '%s'", b.Id)
+				}
+			}
+			return true, nil
+		}
+	}
+	// If the caller is the default task activator, only deactivate tasks that have not been activated by a user.
+	if evergreen.IsSystemActivator(caller) {
+		q[task.ActivatedByKey] = bson.M{"$in": evergreen.SystemActivators}
 	}
 
-	return errors.Wrapf(setTaskActivationForBuilds(buildIDs, active, false, nil, caller),
-		"setting activation for tasks in version '%s'", versionId)
+	tasks, err := task.FindAll(db.Query(q).WithFields(task.IdKey, task.ExecutionKey))
+	if err != nil {
+		return false, errors.Wrap(err, "getting tasks to deactivate")
+	}
+	if len(tasks) > 0 {
+		if err = task.DeactivateTasks(tasks, false, caller); err != nil {
```

Comment:
We should also be updating the build and version status in the deactivate case, right?

Reply:
yes